File: programmers/Lv1/완주하지 못한 선수.py
"""
    lv.1 완주하지 못한 선수
    
    1.  참여자 딕셔너리 생성(값으로는 나타난 횟수를 나타날 때 마다 1씩 증가시키며 저장)
    2.  완주자 목록에 대해, 참여자 딕셔너리의 값을 1씩 뺀다.
    3.  한 명이 완주를 못해 1이 값에 존재하는 키를 반환
"""


def solution(participant, completion):
    answer = {}

    for p in participant:
        answer[p] = answer.get(p, 0) + 1

    for c in completion:
        answer[c] -= 1

    for x in answer:
        if answer[x]:
            return x

File: programmers/Lv1/완주하지 못한 선수.js
/* lv.1 완주하지 못한 선수

1.  완주자 목록을 dict로 만들어 참가자 수와 대조. 초기화 작업 중요 */

function solution(participant, completion) {
  let answer = "";

  const list = {};

  for (let c of completion) {
    if (!list[c]) {
      list[c] = 1;
    } else {
      list[c] += 1;
    }
  }

  for (let p of participant) {
    if (!list[p]) {
      answer = p;
      break;
    }

    list[p] -= 1;

    if (list[p] < 0) {
      answer = p;
      break;
    }
  }

  return answer;
}

const participant = ["marina", "josipa", "nikola", "vinko", "filipa"];
const completion = ["josipa", "filipa", "marina", "nikola"];

console.log(solution(participant, completion));
